File: Python/100.SameTree/solution.py
# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution(object):
    def isSameTree(self, p, q):
        """
        :type p: TreeNode
        :type q: TreeNode
        :rtype: bool
        """
        if p is None and q is None: 
        	return True
        # Both p and q are not None
        elif p and q:
        	return p.val == q.val and self.isSameTree(p.right, q.right) and self.isSameTree(p.left, q.left)
        else:
        	return False

    # Revised
    def isSameTree(self, p, q): 
    	if p and q: 
    		return p.val == q.val and self.isSameTree(p.right, q.right) and self.isSameTree(p.left, q.left)
    	# When comparing with None, use is instead of == as a general rule
    	return p is q

File: Java/100.SameTree/Solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {

	// public boolean isSameTree(TreeNode p, TreeNode q) {
 //        if (p == null && q == null) return true;
 //        if (p == null || q == null) return false;
 //        return (p.val == q.val) && isSameTree(p.left, q.left) && isSameTree(p.right, q.right);
 //    }


	// Level order traversal, not finished yet!
    public boolean isSameTree(TreeNode p, TreeNode q) {
    	Deque<TreeNode> queueA = new LinkedList<>(), queueB = new LinkedList<>();
    	if (p != null) queueA.offer(p);
    	if (q != null) queueB.offer(q);

    	while (!queueA.isEmpty() && !queueB.isEmpty()) {
    		TreeNode nodeA = queueA.poll();
    		TreeNode nodeB = queueB.poll();
            if (nodeA.val != nodeB.val) return false;
            if (!(nodeA.left == null && nodeB.left == null || nodeA.left != null && nodeB.left != null)) {
                return false;
            }
            if (nodeA.left != null) queueA.offer(nodeA.left);
            if (nodeA.right != null) queueA.offer(nodeA.right);
            if (nodeB.left != null) queueB.offer(nodeB.left);
            if (nodeB.right != null) queueB.offer(nodeB.right);
    	}

    	return queueA.isEmpty() && queueB.isEmpty();
    }

    // Better.
    public boolean isSameTree(TreeNode p, TreeNode q) {
        Deque<TreeNode> queueA = new LinkedList<>(), queueB = new LinkedList<>();
        if (p != null) queueA.offer(p);
        if (q != null) queueB.offer(q);

        while (!queueA.isEmpty() && !queueB.isEmpty()) {
            TreeNode nodeA = queueA.poll();
            TreeNode nodeB = queueB.poll();
            if (nodeA.val != nodeB.val) return false;
            
            if (nodeA.left != null) queueA.offer(nodeA.left);
            if (nodeB.left != null) queueB.offer(nodeB.left);

            if (queueA.size() != queueB.size()) return false;

            if (nodeA.right != null) queueA.offer(nodeA.right);
            if (nodeB.right != null) queueB.offer(nodeB.right);

            if (queueA.size() != queueB.size()) return false;
        }

        return queueA.isEmpty() && queueB.isEmpty();
    }

}